README.md
What is this?
=============
A series of four nutrition games: *Name That Food*; *How Many Calories?*; *How Much to Eat?*; and *Is This a Balanced Meal?*

How to use?
===========

Nutrition is not part of the Sugar desktop, but can be added.  Please refer to;

* [How to Get Sugar on sugarlabs.org](https://sugarlabs.org/),
* [How to use Sugar](https://help.sugarlabs.org/),
* [How to use Nutrition](https://help.sugarlabs.org/nutrition.html)

How to upgrade?
===============

On Sugar desktop systems;
* use [My Settings](https://help.sugarlabs.org/my_settings.html), [Software Update](https://help.sugarlabs.org/my_settings.html#software-update), or;
* use Browse to open [activities.sugarlabs.org](https://activities.sugarlabs.org/), search for `Nutrition`, then download.


NEWS
NEWS

16

* Refresh POT file (Jui Pradhan)
* New Translations 
* Remove Stop Accelerator (Aniket Mathur),
* flake8 fixes (Aniket Mathur),
* Port to python 3 (Aniket Mathur),
* Fix PyGIWarnings (Aniket Mathur),
* Update activity.info (Vipul Gupta),
* Add and update README (rdrsadhu, James Cameron)

15

ENHACEMENT:
* New translations

14

* Added repository to activity.info

13

ENHACEMENT:
* New translations

11

ENHACEMENT:
* New translations

9

ENHACEMENTS:
* New translations
* Add summary string

BUG_FIX:
* Fix regression in add custom food code

7

ENHACEMENT:
* Ignacio Rodríguez port to GTK3

6

ENHANCEMENT:
* New food-group game
* New foods from Nicaragua
* Try again 3x when wrong
* Speed up activity launching
* New translations

5

* New translations

4

ENHANCEMENT:
* Switch from Food Pyramid to My Plate

3

ENHANCEMENT:
* Load custom food

2

ENHANCEMENTS:
* Added 4th game: balanced meals
* Add calories to Name game

1

* New project by Walter, Claudia, y Felix



food.py
# -*- coding: utf-8 -*-
from gettext import gettext as _

PYRAMID = [_('sweets'), _('dairy'), _('fruits'), _('meat'), _('grains'),
           _('vegetables')]
# name, calories, food pyramid group, image name
NAME = 0
CALS = 1
GROUP = 2
IMAGE = 3
FOOD_DATA = [
    _('avocado'), 151, 2, 'aguacate.png',
    _('celery'), 19, 5, 'apio.png',
    _('rice'), 343, 4, 'arroz.png',
    _('oats'), 387, 4, 'avena.png',
    _('pumpkin'), 37, 5, 'ayote.png',
    _('banana'), 90, 2, 'banano.png',
    # TRANS: aubergine
    _('eggplant'), 26, 5, 'berenjena.png',
    _('bread'), 69, 4, 'bread.png',
    _('broccoli'), 14, 5, 'brocoli.png',
    _('cocoa'), 343, 4, 'cacao-fruto.png',
    _('cake'), 387, 0, 'cake.png',
    _('barley'), 350, 4, 'cebada.png',
    _('onion'), 40, 5, 'cebolla.png',
    _('pork'), 194, 3, 'cerdo.png',
    _('cherry'), 60, 2, 'cereza.png',
    _('pear squash'), 30, 5, 'chayote.png',
    _('chili pepper'), 23, 5, 'chile.png',
    _('sweet pepper'), 48, 5, 'chiltoma.png',
    _('coconut'), 300, 2, 'coco.png',
    _('cookie'), 68, 0, 'cookie.png',
    _('cream'), 340, 1, 'crema.png',
    _('egg'), 77, 3, 'egg.png',
    _('corn on the cob'), 96, 5, 'elote.png',
    _('strawberry'), 23, 2, 'fresa.png',
    _('guava'), 55, 2, 'guayabas.png',
    _('milk'), 57, 1, 'leche.png',
    _('lettuce'), 19, 5, 'lechuga.png',
    _('lemon'), 30, 2, 'limon.png',
    _('flax seed'), 450, 4, 'linaza.png',
    # _('corn'), 346, 4, 'maiz.png',
    _('mandarin'), 44, 2, 'mandarina.png',
    _('mango'), 73, 2, 'mango.png',
    _('apple'), 65, 2, 'manzana.png',
    _('star fruit'), 63, 2, 'melocoton.png',
    _('melon'), 80, 2, 'melon.png',
    _('orange'), 38, 2, 'naranja.png',
    # TRANS: fruit of the sapodilla tree common to Central America
    _('sapodilla'), 50, 2, 'nispero.png',
    _('potato'), 90, 5, 'papa.png',
    _('papaya'), 25, 2, 'papaya.png',
    _('cucumber'), 12, 5, 'pepino.png',
    _('pear'), 61, 2, 'pera.png',
    _('fish'), 91, 3, 'pez.png',
    _('pineapple'), 55, 2, 'pina.png',
    # TRANS: Green pumpkin squash (source of seeds used in Central American)
    _('green pumpkin'), 16, 5, 'pipian.png',
    # TRANS: a cactus fruit
    _('pitaya'), 36, 2, 'pitaya.png',
    _('plantain'), 86, 5, 'platano.png',
    _('chicken'), 170, 3, 'pierna-de-pollo.png',
    _('cheese'), 7, 1, 'queso.png',
    _('radish'), 16, 5, 'rabano.png',
    _('beet'), 43, 5, 'remolacha.png',
    _('cabbage'), 40, 5, 'repollo.png',
    _('beef'), 180, 3, 'res.png',
    _('rice and beans'), 411, 4, 'rice-and-beans.png',
    _('watermelon'), 47, 2, 'sandia.png',
    # TRANS: A seed used to make horchata (common in Central America)
    _('jicaro'), 30, 4, 'semillas-de-jicaro.png',
    # _('tamale'), 126, 5, 'nacatamal.png',
    _('tomato'), 16, 5, 'tomate.png',
    _('wheat'), 345, 4, 'trigo.png',
    _('grape'), 68, 2, 'uvas.png',
    _('yogurt'), 38, 1, 'yogurt.png',
    _('yucca'), 120, 5, 'yuca.png',
    _('carrot'), 37, 5, 'zanahoria.png',
    # TRANS: a soft, edible fruit native to Central America
    _('sapote'), 52, 2, 'zapote.png',
]


ImageIDActivity.py
#Copyright (c) 2011,12 Walter Bender

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# You should have received a copy of the GNU General Public License
# along with this library; if not, write to the Free Software
# Foundation, 51 Franklin Street, Suite 500 Boston, MA 02110-1335 USA


import gtk

from sugar.activity import activity
from sugar import profile
from sugar.graphics.toolbarbox import ToolbarBox
from sugar.activity.widgets import ActivityToolbarButton
from sugar.activity.widgets import StopButton

from gettext import gettext as _

from game import Game
from toolbar_utils import separator_factory, radio_factory, label_factory

import logging
_logger = logging.getLogger('imageid-activity')


SERVICE = 'org.sugarlabs.ImageIDActivity'
IFACE = SERVICE
LABELS = [_('Match the food to its name.'),
          _('How many calories are there?'),
          _('How much should you eat?')]


class ImageIDActivity(activity.Activity):
    """ Simple reading game based on GCompris ImageID """

    def __init__(self, handle):
        """ Initialize the toolbars and the game board """
        super(ImageIDActivity, self).__init__(handle)

        self.path = activity.get_bundle_path()

        self._setup_toolbars()

        # Create a canvas
        canvas = gtk.DrawingArea()
        canvas.set_size_request(gtk.gdk.screen_width(), \
                                gtk.gdk.screen_height())
        self.set_canvas(canvas)
        canvas.show()
        self.show_all()

        self._game = Game(canvas, parent=self, path=self.path)
        self._game.new_game()

    def _setup_toolbars(self):
        """ Setup the toolbars. """

        self.max_participants = 1  # No collaboration

        toolbox = ToolbarBox()

        # Activity toolbar
        activity_button = ActivityToolbarButton(self)

        toolbox.toolbar.insert(activity_button, 0)
        activity_button.show()

        self.set_toolbar_box(toolbox)
        toolbox.show()
        self.toolbar = toolbox.toolbar

        self.beginner_button = radio_factory(
            'beginner',
            toolbox.toolbar,
            self._level_cb,
            cb_arg=0,
            tooltip=_(LABELS[0]),
            group=None)
        self.intermediate_button = radio_factory(
            'intermediate',
            toolbox.toolbar,
            self._level_cb,
            cb_arg=1,
            tooltip=_(LABELS[1]),
            group=self.beginner_button)
        self.expert_button = radio_factory(
            'expert',
            toolbox.toolbar,
            self._level_cb,
            cb_arg=2,
            tooltip=_(LABELS[2]),
            group=self.beginner_button)

        separator_factory(toolbox.toolbar, False, True)
        self._label = label_factory(toolbox.toolbar, LABELS[0])

        separator_factory(toolbox.toolbar, True, False)
        stop_button = StopButton(self)
        stop_button.props.accelerator = '<Ctrl>q'
        toolbox.toolbar.insert(stop_button, -1)
        stop_button.show()

    def _level_cb(self, button, level):
        ''' Switch between levels '''
        self._game.level = level
        self._label.set_text(LABELS[level])
        self._game.new_game()


siennapierre/frenchFood.py
from gettext import gettext as _

PYRAMID = [_('sucre'), _('laitier'), _('fruits'), _('viande'), _('céréales'),
           _('légumes')]
# name, calories, food pyramid group, image name
NAME = 0
CALS = 1
GROUP = 2
IMAGE = 3
FOOD_DATA = [
_('acovat'), 151, 2, 'aguacate.png',
_('céleri'), 19, 5, 'apio.png',
_('riz'), 343, 4, 'arroz.png',
_('avoine'), 387, 4, 'avena.png',
_('citrouille'), 37, 5, 'ayote.png',
_('banane'), 90, 2, 'banano.png',
_('aubergine'), 26, 5, 'berenjena.png',
_('pain'), 69, 4, 'bread.png',
_('brocoli'), 14, 5, 'brocoli.png',
_('cacao'), 343, 4, 'cacao-fruto.png',
_('gâteau'), 387, 0, 'cake.png',
_('orge'), 350, 4, 'cebada.png',
_('oignon'), 40, 5, 'cebolla.png',
_('porc'), 194, 3, 'cerdo.png',
_('cerise'), 60, 2, 'cereza.png',
_('courge de poire'), 30, 5, 'chayote.png',
_('piment'), 23, 5, 'chile.png',
_('poivron'), 48, 5, 'chiltoma.png',
_('noix de coco'), 300, 2, 'coco.png',
_('biscuit'), 68, 0, 'cookie.png',
_('créme'), 340, 1, 'crema.png',
_('oeuf'), 77, 3, 'egg.png',
_('maïs en épis'), 96, 5, 'elote.png',
_('fraise'), 23, 2, 'fresa.png',
_('goyave'), 55, 2, 'guayabas.png',
_('lait'), 57, 1, 'leche.png',
_('salade'), 19, 5, 'lechuga.png',
_('citron'), 30, 2, 'limon.png',
_('graine de lin'), 450, 4, 'linaza.png',
_('blé'), 346, 4, 'maiz.png',
_('mandarin'), 44, 2, 'mandarina.png',
_('mangue'), 73, 2, 'mango.png',
_('pomme'), 65, 2, 'manzana.png',
_('carambole'), 63, 2, 'melocoton.png',
_('melon'), 80, 2, 'melon.png',
_('orange'), 38, 2, 'naranja.png',
#TRANS: fruit of the sapodilla tree common to Central America
_('sapotille'), 50, 2, 'nispero.png',
_('patate'), 90, 5, 'papa.png',
_('papaye'), 25, 2, 'papaya.png',
_('concombre'), 12, 5, 'pepino.png',
_('paire'), 61, 2, 'pera.png',
_('poisson'), 91, 3, 'pez.png',
_('ananas'), 55, 2, 'pina.png',
#TRANS: Green pumpkin squash (source of seeds used in Central American)
_('citrouille verte'), 16, 5, 'pipian.png',
#TRANS: a cactus fruit
_('pitaya'), 36, 2, 'pitaya.png',
_('banane plantain'), 86, 5, 'platano.png',
_('poulet'), 170, 3, 'pierna-de-pollo.png',
_('fromage'), 07, 1, 'queso.png',
_('radis'), 16, 5, 'rabano.png',
_('betterave'),	43, 5, 'remolacha.png',
_('chou'), 40, 5, 'repollo.png',
_('boeuf'), 180, 3, 'res.png',
_('riz et haricots'), 411, 4, 'rice-and-beans.png',
_('pastéque'), 47, 2, 'sandia.png',
_('tomate'), 16, 5, 'tomate.png',
_('blé'), 345, 4, 'trigo.png',
_('grain de raisin'), 68, 2, 'uvas.png',
_('yaourt'), 38, 1, 'yogurt.png',
_('yucca'), 120, 5, 'yuca.png',
_('carotte'), 37, 5, 'zanahoria.png',
#TRANS: a soft, edible fruit native to Central America
_('sapote'), 52, 2, 'zapote.png',
]


game.py
# -*- coding: utf-8 -*-
# Copyright (c) 2011,12 Walter Bender
# Ported to GTK3:
# Ignacio Rodríguez <ignaciorodriguez@sugarlabs.org>
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# You should have received a copy of the GNU General Public License
# along with this library; if not, write to the Free Software
# Foundation, 51 Franklin Street, Suite 500 Boston, MA 02110-1335 USA

from gi.repository import Gtk, Gdk, GObject, GdkPixbuf
import os
from random import uniform

from gettext import gettext as _

import logging
_logger = logging.getLogger('nutrition-activity')

try:
    from sugar3.graphics import style
    GRID_CELL_SIZE = style.GRID_CELL_SIZE
except ImportError:
    GRID_CELL_SIZE = 0

from food import FOOD_DATA, PYRAMID, NAME, CALS, GROUP, IMAGE

from sprites import Sprites, Sprite

# ChooseMyPlate.gov
MYPLATE = [[PYRAMID[0], 0], [PYRAMID[1], 1], [PYRAMID[2], 2],
           [PYRAMID[3], 2], [PYRAMID[4], 3], [PYRAMID[5], 3]]
QUANT = 1
QUANTITIES = [_('minimum'), _('moderate'), _('more'), _('most')]
BALANCE = [_('balanced'), _('unbalanced')]
NCARDS = 5
FOOD = []


class Game():

    def __init__(self, canvas, parent=None, path=None):
        self._canvas = canvas
        self._parent = parent
        self._parent.show_all()
        self._path = path

        self._canvas.add_events(Gdk.EventMask.BUTTON_PRESS_MASK)
        self._canvas.connect("draw", self.__draw_cb)
        self._canvas.connect("button-press-event", self._button_press_cb)

        self._width = Gdk.Screen.width()
        self._height = Gdk.Screen.height()
        self._scale = self._width / 1200.
        self._target = 0
        self._tries = 0

        self.level = 0

        self._picture_cards = []
        self._small_picture_cards = []
        self.food_cards = []
        self._group_cards = []
        self._quantity_cards = []
        self._balance_cards = []
        self._last_twenty = []
        self._background = None

        # Generate the sprites we'll need...
        self._sprites = Sprites(self._canvas)
        self._background = Sprite(
            self._sprites, 0, 0, GdkPixbuf.Pixbuf.new_from_file_at_size(
                os.path.join(self._path, 'images', 'background.png'),
                self._width, self._height))
        self._background.set_layer(0)
        self._background.type = None
        self._background.hide()

        self.pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_size(
            os.path.join(self._path, 'images', 'word-box.png'),
            int(350 * self._scale), int(100 * self._scale))

        for i in range(int(len(FOOD_DATA) / 4)):
            FOOD.append([FOOD_DATA[i * 4 + NAME], FOOD_DATA[i * 4 + CALS],
                         FOOD_DATA[i * 4 + GROUP], FOOD_DATA[i * 4 + IMAGE]])
            self.food_cards.append(None)
            self._picture_cards.append(None)
            for j in range(6):
                self._small_picture_cards.append(None)
        self.allocate_food(0)

        x = 10
        dx, dy = self.food_cards[0].get_dimensions()

        y = 10
        for i in range(len(MYPLATE)):
            self.word_card_append(self._group_cards, self.pixbuf)
            self._group_cards[-1].type = i
            self._group_cards[-1].set_label(MYPLATE[i][0])
            self._group_cards[-1].move((x, y))
            y += int(dy * 1.25)

        y = 10
        for i in range(len(QUANTITIES)):
            self.word_card_append(self._quantity_cards, self.pixbuf)
            self._quantity_cards[-1].type = i
            self._quantity_cards[-1].set_label(QUANTITIES[i])
            self._quantity_cards[-1].move((x, y))
            y += int(dy * 1.25)

        y = 10
        for i in range(len(BALANCE)):
            self.word_card_append(self._balance_cards, self.pixbuf)
            self._balance_cards[-1].type = i
            self._balance_cards[-1].set_label(BALANCE[i])
            self._balance_cards[-1].move((x, y))
            y += int(dy * 1.25)

        self._smile = Sprite(self._sprites,
                             int(self._width / 4),
                             int(self._height / 4),
                             GdkPixbuf.Pixbuf.new_from_file_at_size(
                                 os.path.join(
                                     self._path,
                                     'images',
                                     'correct.png'),
                                 int(self._width / 2),
                                 int(self._height / 2)))
        self._smile.set_label_attributes(36)
        self._smile.set_margins(10, 0, 10, 0)

        self._frown = Sprite(self._sprites,
                             int(self._width / 4),
                             int(self._height / 4),
                             GdkPixbuf.Pixbuf.new_from_file_at_size(
                                 os.path.join(
                                     self._path,
                                     'images',
                                     'wrong.png'),
                                 int(self._width / 2),
                                 int(self._height / 2)))
        self._frown.set_label_attributes(36)
        self._frown.set_margins(10, 0, 10, 0)

        self.build_food_groups()

        self._all_clear()

    def allocate_food(self, i):
        self.picture_append(os.path.join(self._path, 'images',
                                         FOOD_DATA[i * 4 + IMAGE]), i)
        self.small_picture_append(os.path.join(self._path, 'images',
                                               FOOD_DATA[i * 4 + IMAGE]), i)
        self.word_card_append(self.food_cards, self.pixbuf, i)
        self.food_cards[i].type = i
        self.food_cards[i].set_label(FOOD_DATA[i * 4 + NAME])

    def word_card_append(self, card_list, pixbuf, i=-1):
        if i == -1:
            card_list.append(Sprite(self._sprites, 10, 10, pixbuf))
        else:
            card_list[i] = Sprite(self._sprites, 10, 10, pixbuf)
        card_list[i].set_label_attributes(36)
        card_list[i].set_margins(10, 0, 10, 0)
        card_list[i].hide()

    def picture_append(self, path, i=-1):
        spr = Sprite(
            self._sprites,
            int(self._width / 2.),
            int(self._height / 4.),
            GdkPixbuf.Pixbuf.new_from_file_at_size(
                path, int(self._width / 3.), int(9 * self._width / 12.)))
        if i == -1:
            self._picture_cards.append(spr)
        else:
            self._picture_cards[i] = spr
        self._picture_cards[i].type = 'picture'
        self._picture_cards[i].hide()

    def small_picture_append(self, path, i=-1):
        x = int(self._width / 3.)
        y = int(self._height / 6.)
        pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_size(
            path,
            int(self._width / 6.),
            int(3 * self._width / 8.))
        for j in range(6):  # up to 6 of each card
            if i == -1:
                self._small_picture_cards.append(Sprite(
                    self._sprites, x, y, pixbuf))
                self._small_picture_cards[-1].type = 'picture'
                self._small_picture_cards[-1].hide()
            else:
                self._small_picture_cards[i * 6 + j] = Sprite(
                    self._sprites, x, y, pixbuf)
                self._small_picture_cards[i * 6 + j].type = 'picture'
                self._small_picture_cards[i * 6 + j].hide()
            x += int(self._width / 6.)
            if j == 2:
                x = int(self._width / 3.)
                y += int(3 * self._width / 16.)

    def _all_clear(self):
        ''' Things to reinitialize when starting up a new game. '''
        for p in self._picture_cards:
            if p is not None:
                p.hide()
        for p in self._small_picture_cards:
            if p is not None:
                p.hide()
        for i, w in enumerate(self.food_cards):
            if w is not None:
                w.set_label_color('black')
                w.set_label(FOOD[i][NAME])
                w.hide()
        for i, w in enumerate(self._group_cards):
            w.set_label_color('black')
            w.set_label(MYPLATE[i][0])
            w.hide()
        for i, w in enumerate(self._quantity_cards):
            w.set_label_color('black')
            w.set_label(QUANTITIES[i])
            w.hide()
        for i, w in enumerate(self._balance_cards):
            w.set_label_color('black')
            w.set_label(BALANCE[i])
            w.hide()
        self._smile.hide()
        self._frown.hide()

        self._background.set_layer(1)

    def build_food_groups(self):
        self._my_plate = [[], [], [], []]
        for i, food in enumerate(FOOD):
            self._my_plate[MYPLATE[food[GROUP]][QUANT]].append(i)

    def new_game(self):
        ''' Start a new game. '''
        games = {0: self._name_that_food, 1: self._name_that_food_group,
                 2: self._compare_calories, 3: self._how_much_to_eat,
                 4: self._balanced_meal}
        self._all_clear()

        games[self.level]()

        self._frown.set_label('')
        self._smile.set_label('')
        self._tries = 0

    def _name_that_food(self):
        ''' Choose food cards and one matching food picture '''
        x = 10
        y = 10
        dx, dy = self.food_cards[0].get_dimensions()

        # Select some cards
        word_list = []
        for i in range(NCARDS):
            j = int(uniform(0, len(FOOD)))
            while j in word_list:
                j = int(uniform(0, len(FOOD)))
            word_list.append(j)

        # Show the word cards from the list
        for i in word_list:
            if self.food_cards[i] is None:
                self.allocate_food(i)
            self.food_cards[i].set_layer(100)
            self.food_cards[i].move((x, y))
            y += int(dy * 1.25)

        # Choose a random food image from the list and show it.
        self._target = self.food_cards[
            word_list[int(uniform(0, NCARDS))]].type
        while self._target in self._last_twenty:
            self._target = self.food_cards[
                word_list[int(uniform(0, NCARDS))]].type
        self._last_twenty.append(self._target)
        if len(self._last_twenty) > 20:
            self._last_twenty.remove(self._last_twenty[0])

        self._picture_cards[self._target].set_layer(100)

    def _name_that_food_group(self):
        ''' Show group cards and one food picture '''
        for i in range(len(MYPLATE)):
            self._group_cards[i].set_layer(100)

        # Choose a random food image and show it.
        self._target = int(uniform(0, len(FOOD)))
        if self.food_cards[self._target] is None:
            self.allocate_food(self._target)
        self._picture_cards[self._target].set_layer(100)

    def _compare_calories(self):
        ''' Choose food cards and compare the calories '''
        x = 10
        y = 10
        dx, dy = self.food_cards[0].get_dimensions()

        # Select some cards
        word_list = []
        for i in range(6):
            j = int(uniform(0, len(FOOD)))
            while j in word_list:
                j = int(uniform(0, len(FOOD)))
            word_list.append(j)
            if self.food_cards[j] is None:
                self.allocate_food(j)

        # Show the word cards from the list
        for i in word_list:
            self.food_cards[i].set_layer(100)
            self.food_cards[i].move((x, y))
            y += int(dy * 1.25)

        # Show food images
        self._target = word_list[0]
        for i in range(5):
            if FOOD[word_list[i + 1]][CALS] > FOOD[self._target][CALS]:
                self._target = word_list[i + 1]
        self._small_picture_cards[word_list[0] * 6].set_layer(100)
        self._small_picture_cards[word_list[1] * 6 + 1].set_layer(100)
        self._small_picture_cards[word_list[2] * 6 + 2].set_layer(100)
        self._small_picture_cards[word_list[3] * 6 + 3].set_layer(100)
        self._small_picture_cards[word_list[4] * 6 + 4].set_layer(100)
        self._small_picture_cards[word_list[5] * 6 + 5].set_layer(100)

    def _how_much_to_eat(self):
        ''' Show quantity cards and one food picture '''
        for i in range(len(QUANTITIES)):
            self._quantity_cards[i].set_layer(100)

        # Choose a random image from the list and show it.
        self._target = int(uniform(0, len(FOOD)))
        if self.food_cards[self._target] is None:
            self.allocate_food(self._target)
        self._picture_cards[self._target].set_layer(100)

    def _balanced_meal(self):
        ''' A well-balanced meal '''
        for i in range(2):
            self._balance_cards[i].set_layer(100)

        # Determine how many foods from each group
        n = [0, 0, 0, 0]
        n[0] = int(uniform(0, 2.5))
        n[1] = int(uniform(0, 3 - n[0]))
        n[2] = 3 - n[0] - n[1]
        n[3] = 6 - n[0] - n[1] - n[2]

        # Fill a plate with foods from different groups
        meal = []
        for i in range(n[0]):  # Sweets
            j = int(uniform(0, len(self._my_plate[0])))
            meal.append(self._my_plate[0][j])
        for i in range(n[1]):  # Dairy
            j = int(uniform(0, len(self._my_plate[1])))
            meal.append(self._my_plate[1][j])
        for i in range(n[2]):  # Protein and Fruits
            j = int(uniform(0, len(self._my_plate[2])))
            meal.append(self._my_plate[2][j])
        for i in range(n[3]):  # Veggies and Grains
            j = int(uniform(0, len(self._my_plate[3])))
            meal.append(self._my_plate[3][j])

        if n[0] < 2 and n[1] < 2 and n[2] < n[3]:
            self._target = 0  # Balanced meal
        else:
            self._target = 1

        for i in range(6):
            if self.food_cards[meal[i]] is None:
                self.allocate_food(meal[i])
        # Randomly position small cards
        self._small_picture_cards[meal[3] * 6].set_layer(100)
        self._small_picture_cards[meal[4] * 6 + 1].set_layer(100)
        self._small_picture_cards[meal[1] * 6 + 2].set_layer(100)
        self._small_picture_cards[meal[2] * 6 + 3].set_layer(100)
        self._small_picture_cards[meal[5] * 6 + 4].set_layer(100)
        self._small_picture_cards[meal[0] * 6 + 5].set_layer(100)

    def _button_press_cb(self, win, event):
        win.grab_focus()
        x, y = list(map(int, event.get_coords()))
        spr = self._sprites.find_sprite((x, y))
        if spr is None:
            return
        # We only care about clicks on word cards
        if not isinstance(spr.type, int):
            return

        # Which card was clicked? Set its label to red.
        spr.set_label_color('red')
        label = spr.labels[0]
        spr.set_label(label)

        if self.level == 0:
            if spr.type == self._target:
                self._smile.set_layer(200)
                self._tries = 3
            else:
                self._frown.set_layer(200)
                self._tries += 1
            if self._tries == 3:
                self.food_cards[self._target].set_label_color('blue')
                label = self.food_cards[self._target].labels[0]
                self.food_cards[self._target].set_label(label)
        elif self.level == 1:
            i = FOOD[self._target][GROUP]
            if spr.type == i:
                self._smile.set_layer(200)
                self._tries = 3
            else:
                self._frown.set_layer(200)
                self._tries += 1
            if self._tries == 3:
                self._group_cards[i].set_label_color('blue')
                label = self._group_cards[i].labels[0]
                self._group_cards[i].set_label(label)
        elif self.level == 2:
            if spr.type == self._target:
                self._smile.set_layer(200)
                self._tries = 3
            else:
                self._frown.set_layer(200)
                self._tries += 1
            if self._tries == 3:
                self.food_cards[self._target].set_label_color('blue')
                label = self.food_cards[self._target].labels[0]
                self.food_cards[self._target].set_label(label)
        elif self.level == 3:
            i = MYPLATE[FOOD[self._target][GROUP]][QUANT]
            if spr.type == i:
                self._smile.set_layer(200)
                self._tries = 3
            else:
                self._frown.set_layer(200)
                self._tries += 1
            if self._tries == 3:
                self._quantity_cards[i].set_label_color('blue')
                label = self._quantity_cards[i].labels[0]
                self._quantity_cards[i].set_label(label)
        elif self.level == 4:
            if self._target == spr.type:
                self._smile.set_layer(200)
                self._tries = 3
            else:
                self._frown.set_layer(200)
                self._tries += 1
            if self._tries == 3:
                self._balance_cards[self._target].set_label_color('blue')
                label = self._balance_cards[self._target].labels[0]
                self._balance_cards[self._target].set_label(label)
        else:
            _logger.debug('unknown play level %d' % (self.level))

        # Play again
        if self._tries == 3:
            GObject.timeout_add(2000, self.new_game)
        else:
            GObject.timeout_add(1000, self._reset_game)
        return True

    def _reset_game(self):
        self._frown.hide()
        if self.level in [0, 2]:
            for i, w in enumerate(self.food_cards):
                if w is not None:
                    w.set_label_color('black')
                    w.set_label(FOOD[i][NAME])
        elif self.level == 1:
            for i, w in enumerate(self._group_cards):
                w.set_label_color('black')
                w.set_label(MYPLATE[i][0])
        elif self.level == 3:
            for i, w in enumerate(self._quantity_cards):
                w.set_label_color('black')
                w.set_label(QUANTITIES[i])
        elif self.level == 4:
            for i, w in enumerate(self._balance_cards):
                w.set_label_color('black')
                w.set_label(BALANCE[i])

    def __draw_cb(self, canvas, cr):
        self._sprites.redraw_sprites(cr=cr)

    def do_expose_event(self, event):
        ''' Handle the expose-event by drawing '''
        # Restrict Cairo to the exposed area
        cr = self._canvas.window.cairo_create()
        cr.rectangle(event.area.x, event.area.y,
                     event.area.width, event.area.height)
        cr.clip()
        # Refresh sprite list
        self._sprites.redraw_sprites(cr=cr)

    def _destroy_cb(self, win, event):
        Gtk.main_quit()
